packages/@lwc/engine-core/src/framework/utils.ts

```diff
@@ -98,3 +100,20 @@ export function flattenStylesheets(stylesheets: TemplateStylesheetFactories): St
     }
     return list;
 }
+
+// Set a ref (lwc:ref) on a VM, from a template API
+export function setRefVNode(vm: VM, ref: string | undefined, vnode: VBaseElement) {
+    if (!isUndefined(ref)) {
+        // Create the refVNodes object on-demand. This ensures that templates with no refs
+        // don't pay the perf tax of creating objects unnecessarily.
+        const refVNodes: RefVNodes = isNull(vm.refVNodes)
+            ? (vm.refVNodes = create(null))
+            : vm.refVNodes;
+
+        // In cases of conflict (two elements with the same ref), prefer, the last one,
+        // in depth-first traversal order.
+        if (!(ref in refVNodes) || refVNodes[ref].key < vnode.key) {
```

Comment:
the `in` operator is always problematic for me...  e.g.: `'__proto__' in {}`

Reply:
In that case, maybe we're better off with a `Map`. Otherwise users could use the string `constructor` or `__proto__` as the ref key and get something weird.

Reply:
Actually I just realized we already have this problem with `this.refs` anyway, since it's a plain object. So there's little point in converting one to `Map`.

And also, it should not matter in this case because the `refVNodes` object was created with `Object.create(null)`. Or is there something I'm missing?